[retry][dynamo][numpy] Add support for np.dtype

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105034

Original PR: #103546

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `str` and then feed into the corresponding `torch_np` method. The assumption here is that all `torch_np` methods that are taking `dtype` kwarg will be able to also take `str` as `dtype`. This assumption stands for `numpy`.

For previous example, we convert `np.float64` to `"float64"` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

larryliu0820 committed:
[retry][dynamo][numpy] Add support for np.dtype

Original PR: #103546

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `str` and then feed into the corresponding `torch_np` method. The assumption here is that all `torch_np` methods that are taking `dtype` kwarg will be able to also take `str` as `dtype`. This assumption stands for `numpy`.

For previous example, we convert `np.float64` to `"float64"` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.

[ghstack-poisoned]

ezyang commented:
I can stampy but @voznesenskym can you check the fix for ci fialure

voznesenskym commented on test/dynamo/test_misc.py:
```diff
@@ -1278,21 +1278,59 @@ def fn(x, y):
             ref = fn(x, y)
             res = opt_fn(x, y)
             self.assertTrue(same(ref, res))
+        # we expect fn to have 2 frames
```
why? Where's the graph break?

larryliu0820 replied:
Passing numpy ndarray to `torch.tensor` is causing a graph break and it is the expected behavior.

voznesenskym commented on test/dynamo/test_misc.py:
```diff
@@ -1278,21 +1278,59 @@ def fn(x, y):
+            x = torch.randint(100, (1,))
+            ref = mandelbrot_numpy(x.item())
+            res = opt_fn(x.item())
+            self.assertTrue(same(ref, res))
```
How many frames is this in compile counter?

larryliu0820 replied:
nopython=True so 1?

voznesenskym replied:
Or 0 ;) No frames because no compiles will still pass with nopython=True AND will give a correct response on result comparison.

larryliu0820 replied:
actually it creates multiple graphs because of the for loop. Not too sure what happens here

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -815,8 +815,12 @@ def call_function(
+        # lookup method name in torch_np. Things like np.dtype(float) are not supported yet.
+        if self.value.__name__ == "dtype":
+            unimplemented(
+                f"numpy dtype function is not supported yet. Got {self.value}."
```
Let's not f string the `value` as that could be a huge array, right?

larryliu0820 replied:
type(self.value)?

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -815,8 +815,12 @@ def call_function(
         from .tensor import NumpyNdarrayVariable
 
         options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
-        # lookup method name in torch_np
```
The class doc is "Wrapper around `numpy.*` for better error messages."

But we are clearly doing more than that - time to change it.

Also, can you assert that self.value has the expected type in __init__?

I know this is not perfectly related to your PR, but you are adding real logic here, so its time to bring this class up to par.

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -815,8 +815,12 @@ def call_function(
+            unimplemented(
+                f"numpy dtype function is not supported yet. Got {self.value}."
+            )
+        elif hasattr(torch_np, self.value.__name__):
```
Assert that there is no custom attribute access implementation here. We have utils for this. 

The old code didn't do it, and could potentially cause us to invoke arbitrary user code during the hasattr.

voznesenskym replied:
`object_has_getattribute` `get_custom_getattr`

larryliu0820 replied:
Found it

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -815,8 +815,12 @@ def call_function(
+            )
+        elif hasattr(torch_np, self.value.__name__):
             func = getattr(torch_np, self.value.__name__)
             return wrap_fx_proxy_cls(
```
The contract here is sus to me? any function on any numpy variable always returns a NumpyNdarrayVariable we pack with tensor types? This is totally out of scope for your PR, something I need to follow up on.

larryliu0820 replied:
Yeah you are right, it's not guaranteed that any function on any numpy variable always returns a NumpyNdarrayVariable. This is definitely still WIP.

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -847,6 +851,22 @@ def python_type(self):
 
+    def as_proxy(self):
+        # this handles numpy dtype attribute such as np.float32.
+        if isinstance(self.value, type) and config.numpy_ndarray_as_tensor:
```
You should leave a TODO here to split NumpyVariable into NumpyVariables for instances/objects and NumpyVariable for types, this is unusual.

voznesenskym reviewed (approved):
PR contents are good, ty for doing this. The dynamo<>numpy code has some sus edges I need to audit.

larryliu0820 committed:
Update on "[retry][dynamo][numpy] Add support for np.dtype"


Original PR: #103546

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `str` and then feed into the corresponding `torch_np` method. The assumption here is that all `torch_np` methods that are taking `dtype` kwarg will be able to also take `str` as `dtype`. This assumption stands for `numpy`.

For previous example, we convert `np.float64` to `"float64"` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

larryliu0820 committed:
Update on "[retry][dynamo][numpy] Add support for np.dtype"


Original PR: #103546

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `str` and then feed into the corresponding `torch_np` method. The assumption here is that all `torch_np` methods that are taking `dtype` kwarg will be able to also take `str` as `dtype`. This assumption stands for `numpy`.

For previous example, we convert `np.float64` to `"float64"` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge
